lib/internal/fs/promises.js

```diff
@@ -663,19 +666,55 @@ async function writeFile(path, data, options) {
   options = getOptions(options, { encoding: 'utf8', mode: 0o666, flag: 'w' });
   const flag = options.flag || 'w';
 
-  if (!isArrayBufferView(data)) {
-    validateStringAfterArrayBufferView(data, 'data');
-    data = Buffer.from(data, options.encoding || 'utf8');
-  }
+  if (isIterable(data)) {
+    if (options.signal?.aborted) {
+      throw lazyDOMException('The operation was aborted', 'AbortError');
+    }
```

Comment:
Optionally we could have the abort signal check outside the if block so it shortcuts the second `open` too.
```suggestion
  if (options.signal?.aborted) {
    throw lazyDOMException('The operation was aborted', 'AbortError');
  }

  if (isIterable(data)) {
```